de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/tagging/dialog/TagDialog.java

```diff
@@ -0,0 +1,99 @@
+package de.tu_bs.cs.isf.e4cf.parts.project_explorer.tagging.dialog;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+import org.eclipse.e4.core.contexts.IEclipseContext;
+
+import de.tu_bs.cs.isf.e4cf.core.file_structure.FileTreeElement;
+import de.tu_bs.cs.isf.e4cf.core.util.ServiceContainer;
+import de.tu_bs.cs.isf.e4cf.parts.project_explorer.tagging.Tag;
+import de.tu_bs.cs.isf.e4cf.parts.project_explorer.tagging.TagService;
+import javafx.scene.control.Alert;
+import javafx.scene.control.Alert.AlertType;
+import javafx.scene.control.ButtonType;
+import javafx.scene.control.DialogPane;
+
+/**
+ * A dialog that display the page in which the user can add custom tags to
+ * FileTreeElements.
+ */
+public class TagDialog {
+
+	private Alert alert;
+	private TagPage tagPage;
+	// indicates whether a file has been moved.
+	private TagService tagService;
+	private List<FileTreeElement> selectedElements;
+	private ServiceContainer services;
+
+	List<Tag> previouslySelected = new ArrayList<Tag>();
+
+	public TagDialog(IEclipseContext context, TagService tagService, ServiceContainer services,
+			List<FileTreeElement> selectedElements) {
+		alert = new Alert(AlertType.NONE);
+		alert.setTitle("Manage Tags");
+
+		this.tagService = tagService;
+		this.services = services;
+
+		this.selectedElements = selectedElements;
+
+		previouslySelected = getInitialSelectedTags();
+
+		final DialogPane pane = alert.getDialogPane();
+		pane.getButtonTypes().addAll(ButtonType.OK, ButtonType.CANCEL);
+
+		this.tagPage = new TagPage(context, tagService, previouslySelected);
+		pane.setContent(tagPage.createControl());
+
+	}
+
+	/** show the dialog and wait for the users input */
+	public void open() {
+		alert.showAndWait().filter(response -> response == ButtonType.OK).ifPresent(response -> {

```

Comment:
Make sure that pressing cancel does cancel actions taken in the wizard (creating, deleting  tags etc)

Reply:
In my opinion there are two approaches

1. determine what has changed and apply those changes to the tagService by calling addTag when a new tag has been added and delete when a tag has been deleted
2. my favorite approach :D just work directly one the available tag list, since all the methods of the service do the same and we even get the correct reference when calling stuff like getavaiable tags. atm i don't see the downside of that.